### PR TITLE
Make the 32Blit tool autodetect the COM port of the 32Blit (on Windows)

### PR DESCRIPTION
**Pull request by lenardg**
The Windows version was just getting the first USB COM port. Since there can be many (I had two others BEFORE the 32Blit), this is not a good way to work.

Added code to actually detect the COM port based on the PID and VID of the USB device. (I found this method of detecting device on the internet, adapted it to the 32Blit tool).

### CODE REVIEW

**Comment by lenardg** [comment]
Sorry about the ugly code, I based it off from somewhere but needed to make changes to work in the 32Blit tool and then it ended up being this mix of C and C++ style blaah.

**Comment by Daft-Freak** [comment]
Uh oh, looks like mingw is missing `DEVPROPTYPE`.

**Comment by Daft-Freak** [comment]
CMake needs something like in `tools/src/CMakeLists.txt`:
```
if(WIN32)
    target_link_libraries(32Blit setupapi)
endif()
```

MinGW may just need `DEVPROPTYPE` swapped for `DWORD`.

Should hopefully get it building...

**Comment by lenardg** [comment]
I will check it out tomorrow, and try to fix it.

**Comment by Daft-Freak** [comment]
Hmm, since this compiles fine with my local mingw, updating the travis build to 20.04 will probably also fix it. (Similar to [the recent change for STM32](https://github.com/pimoroni/32blit-beta/commit/e1e72901930f9fcd0467ac21a4c154171abd7974)).

File: tools/src/32Blit.cpp [hunk 1]
```diff
@@ -162,22 +163,90 @@ bool HandleRX()
...
-  return "";
+    HDEVINFO DeviceInfoSet;
+    DWORD DeviceIndex = 0;
+    SP_DEVINFO_DATA DeviceInfoData;
+    PCSTR DevEnum = "USB";
+    TCHAR ExpectedDeviceId[80] = { 0 }; //Store hardware id
+    BYTE szBuffer[1024] = { 0 };
+    DEVPROPTYPE ulPropertyType;
+    DWORD dwSize = 0;
+    DWORD Error = 0;
```
Comment (Daft-Freak) [inline tools/src/32Blit.cpp:178]: This isn't used. (apart from setting it below).

File: tools/src/32Blit.cpp [hunk 2]
```diff
@@ -162,22 +163,90 @@ bool HandleRX()
...
-    if (QueryDosDevice(portName.c_str(), targetPath, 1024) && std::string(targetPath).find("USB") != std::string::npos)
-      return portName;
-  }
-
-  return "";
+    HDEVINFO DeviceInfoSet;
+    DWORD DeviceIndex = 0;
+    SP_DEVINFO_DATA DeviceInfoData;
+    PCSTR DevEnum = "USB";
+    TCHAR ExpectedDeviceId[80] = { 0 }; //Store hardware id
```
Comment (Daft-Freak) [inline tools/src/32Blit.cpp:174]: `const char *` ... `= "VID_0483&PID_5740"` and avoid the strcpy/cast later?

**Comment by Gadgetoid** [comment]
@lenardg thanks, this could be super handy! The Python flash tool is still lagging behind this one.

Did you ever manage to look into the suggested tweaks & get it building?

**Comment by lenardg** [comment]
I have been pretty snowed under lately from work, but I really really would want to get back into 32Blit. I will try to look into these this weekend, if that is an ok timetable for you guys?

**Comment by Gadgetoid** [comment]
In your own time. I didn't want to push you into doing anything, but rather try and clean house with our outstanding PRs. I prefer to get things merged to achknowledge people's work if I can. Plus this is a useful change.

But yes, this weekend or whenever suits you. No pressure. Honest!

**Comment by Gadgetoid** [comment]
This might need a bump to kick off the integration tests post The Great Travis CI Exodus. You can push an empty comit or something for that, IIRC!

**Comment by Gadgetoid** [comment]
Okay I'm closing this because the tools have been ripped out of this repository to here - https://github.com/32blit/tools-cpp

By and large the Python tool completely replaces this, but I'll leave the door open for PRs.

**Pull request closed without merging**